### PR TITLE
Refactor nix and make pollinate consumable

### PR DESCRIPTION
**Pull request by ulrikstrid**
Refactored nix to follow approximately what I did in `json-logs-formatter` which exposes a overlay with pollinate and exposes pollinate under packages for each OCaml version that we support.

Also added `public_name` to main `pollinate` library and exposed the `Inbox` module in `Node`.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -14,31 +14,63 @@
...
+            });
+          };
 
-          defaultPackage = myPkgs.pollinate;
+          packages = pkgs.lib.mapAttrs (version: ocamlPackages:
+            (pkgs.callPackage ./nix {
+              inherit ocamlPackages;
+              doCheck = true;
+            })) (builtins.removeAttrs pkgs.ocaml-ng [
+              "overrideDerivation"
```
Comment (gsebil08) [inline flake.nix:42]: Does it mean "we want to remove everything except ocaml 4.13"?
Isn't there any more beautiful way to do it?

File: lib/node/node.ml [hunk 2]
```diff
@@ -4,6 +4,7 @@ open Types
 module Message = Message
 module Client = Client
 module Failure_detector = Failure_detector
+module Inbox = Inbox
```
Comment (gsebil08) [inline lib/node/node.ml:7]: Why?

File: nix/overlay.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
+final: prev: {
+  ocaml-ng = (prev.lib.mapAttrs (_: ocamlVersion:
+    ocamlVersion.overrideScope' (oself: osuper: {
+      pollinate = (final.callPackage ./default.nix {
+        ocamlPackages = osuper;
+        doCheck = true;
+      });
+    })) (builtins.removeAttrs prev.ocaml-ng [
```
Comment (gsebil08) [inline nix/overlay.nix:8]: Looks like it is exactly the same than in `flake.nix`, can we make it using the same to avoid duplication?

**Review by gsebil08 — COMMENTED** [review]
Thanks for this, it looks great.
I am wondering if we should only make it OCaml 5.00 compatible... But not sure...

File: lib/node/node.ml (hunk 2 above)
Comment (ulrikstrid) [reply lib/node/node.ml:7]: It's used in the deku integration branch i tested against, of this is wrong we should revert and fix downstream

File: nix/overlay.nix (hunk 3 above)
Comment (ulrikstrid) [reply nix/overlay.nix:8]: Yes, but I was lazy

File: flake.nix (hunk 1 above)
Comment (ulrikstrid) [reply flake.nix:42]: Yes, we should invert it

**Comment by ulrikstrid** [comment]
If we only want ocaml 5 we can simplify this quite a bit

**Review by gsebil08 — APPROVED** [review]

**Pull request merged**